### PR TITLE
speedup and fix wallet-hd functional test for CA

### PR DESCRIPTION
**Pull request by instagibbs**
rough backport of https://github.com/bitcoin/bitcoin/pull/12947/

### CODE REVIEW

**Comment by instagibbs** [comment]
tests are passing in reality. Stupid push test.

File: qa/rpc-tests/wallet-hd.py [hunk 1]
```diff
@@ -30,13 +29,18 @@ def setup_network(self):
     def run_test (self):
         tmpdir = self.options.tmpdir
 
+        # self-send all the op_true funds
+        self.nodes[0].sendtoaddress(self.nodes[0].getnewaddress(), 21000000, "", "", True)
+        self.nodes[0].generate(100)
```
Comment (jtimon) [inline qa/rpc-tests/wallet-hd.py]: Is this generate needed? there's another generate(101) below.

File: qa/rpc-tests/wallet-hd.py [hunk 2]
```diff
@@ -56,20 +60,19 @@ def run_test (self):
...
         os.remove(self.options.tmpdir + "/node1/elementsregtest/wallet.dat")
         shutil.copyfile(tmpdir + "/hd.bak", tmpdir + "/node1/elementsregtest/wallet.dat")
         self.nodes[1] = start_node(1, self.options.tmpdir, self.node_args[1])
-        #connect_nodes_bi(self.nodes, 0, 1)
+        connect_nodes_bi(self.nodes, 0, 1)
 
         # Assert that derivation is deterministic
         hd_add_2 = None
-        for _ in range(num_hd_adds):
+        for _ in range(NUM_HD_ADDS):
```
Comment (jtimon) [inline qa/rpc-tests/wallet-hd.py]: Can we s/_/i/ like in https://github.com/bitcoin/bitcoin/commit/979f59850c72624137d25f80be4188c3ba6b5fa0 ?

**Comment by jtimon** [comment]
utACK modulo nits

**Comment by instagibbs** [comment]
Pushed nit fixes

**Comment by jtimon** [comment]
previous:
```
TEST                           | PASSED | DURATION

mempool_resurrect_test.py      | True   | 3 s
mempool_spendcoinbase.py       | True   | 4 s
multi_rpc.py                   | True   | 5 s
p2p-fullblocktest.py           | True   | 5 s
signrawtransactions.py         | True   | 6 s
decodescript.py                | True   | 6 s
p2p-mempool.py                 | True   | 7 s
prioritise_transaction.py      | True   | 7 s
abandonconflict.py             | True   | 8 s
mempool_reorg.py               | True   | 8 s
sendheaders.py                 | True   | 9 s
invalidblockrequest.py         | True   | 9 s
invalidtxrequest.py            | True   | 10 s
rpc_getblockstats.py           | True   | 10 s
keypool.py                     | True   | 11 s
httpbasics.py                  | True   | 11 s
signmessages.py                | True   | 12 s
rpcnamedargs.py                | True   | 12 s
p2p-leaktests.py               | True   | 13 s
segwit.py                      | True   | 13 s
rest.py                        | True   | 14 s
txn_doublespend.py --mineblock | True   | 15 s
txn_clone.py                   | True   | 15 s
proxy_test.py                  | True   | 16 s
zmq_test.py                    | True   | 16 s
merkle_blocks.py               | True   | 17 s
fundrawtransaction.py          | True   | 17 s
wallet.py                      | True   | 18 s
reindex.py                     | True   | 18 s
preciousblock.py               | True   | 19 s
replace-by-fee.py              | True   | 19 s
importmulti.py                 | True   | 23 s
zapwallettxes.py               | True   | 23 s
nodehandling.py                | True   | 25 s
receivedby.py                  | True   | 27 s
rawtransactions.py             | True   | 27 s
listtransactions.py            | True   | 29 s
getchaintips.py                | True   | 39 s
listsinceblock.py              | True   | 43 s
confidential_transactions.py   | True   | 49 s
wallet-hd.py                   | True   | 183 s

ALL                            | True   | 821 s (accumulated)
```
After thus change:
```
TEST                           | PASSED | DURATION

mempool_spendcoinbase.py       | True   | 3 s
mempool_resurrect_test.py      | True   | 4 s
rpcnamedargs.py                | True   | 4 s
p2p-fullblocktest.py           | True   | 5 s
multi_rpc.py                   | True   | 5 s
signrawtransactions.py         | True   | 6 s
decodescript.py                | True   | 6 s
p2p-mempool.py                 | True   | 7 s
abandonconflict.py             | True   | 7 s
mempool_reorg.py               | True   | 8 s
sendheaders.py                 | True   | 8 s
prioritise_transaction.py      | True   | 9 s
invalidblockrequest.py         | True   | 9 s
invalidtxrequest.py            | True   | 10 s
keypool.py                     | True   | 10 s
httpbasics.py                  | True   | 11 s
signmessages.py                | True   | 11 s
p2p-leaktests.py               | True   | 12 s
segwit.py                      | True   | 13 s
rest.py                        | True   | 14 s
txn_doublespend.py --mineblock | True   | 14 s
txn_clone.py                   | True   | 15 s
proxy_test.py                  | True   | 15 s
zmq_test.py                    | True   | 16 s
wallet.py                      | True   | 16 s
fundrawtransaction.py          | True   | 17 s
merkle_blocks.py               | True   | 17 s
reindex.py                     | True   | 18 s
preciousblock.py               | True   | 18 s
replace-by-fee.py              | True   | 19 s
importmulti.py                 | True   | 21 s
rawtransactions.py             | True   | 22 s
wallet-hd.py                   | True   | 23 s
zapwallettxes.py               | True   | 24 s
nodehandling.py                | True   | 24 s
listtransactions.py            | True   | 25 s
receivedby.py                  | True   | 29 s
getchaintips.py                | True   | 38 s
confidential_transactions.py   | True   | 41 s
listsinceblock.py              | True   | 41 s

ALL                            | True   | 615 s (accumulated)
```

**Pull request merged**